README


This archive contains the distribution PICA-Modification,
version 0.15:

  Idempotent modifications of identified PICA+ records

This software is copyright (c) 2012 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
2012-09-26  0.15
    * Made modifications idempotent
    * Fixed dependencies

2012-09-25  0.14
    * First release at CPAN


lib/PICA/Modification.pm
package PICA::Modification;
{
  $PICA::Modification::VERSION = '0.15';
}
#ABSTRACT: Idempotent modifications of identified PICA+ records

use strict;
use warnings;
use v5.10;

use parent 'Exporter';

use PICA::Record 0.584;
use Scalar::Util qw(blessed);

our @ATTRIBUTES = qw(id iln epn del add);

sub new {
	my $class = shift;
	my $attributes = @_ % 2 ? (blessed $_[0] ? $_[0]->attributes : $_[0]) : {@_};

    no strict 'refs';
    my $self = bless {
		map { $_ => $attributes->{$_} } @{ $class.'::ATTRIBUTES' }
	}, $class;

	$self->check;
}

sub attributes {
	my $self = shift;

    no strict 'refs';
	return { map { $_ => $self->{$_} } @{ ref($self).'::ATTRIBUTES' } };
}


sub check {
	my $self = shift;

	$self->{errors} = { };

	foreach my $attr (qw(id iln epn del add)) {
		my $value = $self->{$attr} // '';
	    $value =~ s/^\s+|\s+$//g;
		$self->{$attr} = $value;
	}

	$self->{ppn} = '';
	$self->{dbkey} = '';
    if ($self->{id} =~ /^(([a-z]([a-z0-9-]?[a-z0-9]+))*):ppn:(\d+\d*[Xx]?)$/) {
        $self->{ppn}   = uc($4) if defined $4;
        $self->{dbkey} = lc($1) if defined $1;
    } elsif ($self->{id} eq '') {
        $self->error( id => 'missing record identifier' );
    } else {
        $self->error( id => 'malformed record identifier' );
    }

    $self->error( iln => "malformed ILN" ) unless $self->{iln} =~ /^\d*$/;
    $self->error( epn => "malformed EPN" ) unless $self->{epn} =~ /^\d*$/;

    my %must_delete;

    if ($self->{add}) {
        my $pica = eval { PICA::Record->new( $self->{add} ) };
        if ($pica) {
			$self->error( iln => 'missing ILN for add' )
				if !$self->{iln} and $pica->field(qr/^1/);
			$self->error( epn => 'missing EPN for add' )
				if !$self->{epn} and $pica->field(qr/^2/);
            $pica->sort;
            foreach ($pica->fields) {
                my $tag = $_->tag;
                # TODO: remove occurrence from level 2 tags
                $must_delete{$tag} = 1;
            }
	    	$self->{add} = "$pica";
			chomp $self->{add};
        } else {
            $self->error( add => "malformed fields to add" );
        }
    }

	my @del = grep { $_ !~ /^\s*$/ } split(/\s*,\s*/, $self->{del});

	$self->error( del => 'malformed fields to remove' )
        if grep { $_ !~  qr{^[012]\d\d[A-Z@](/\d\d)?$} } @del;

	$self->error( epn => 'missing EPN for remove' )
		if !$self->{epn} and grep { /^2/ } @del;

	$self->error( iln => 'missing ILN for remove' )
		if !$self->{iln} and grep { /^1/ } @del;

    delete $must_delete{$_} for @del;
    if (%must_delete) {
        $self->error( del => 'fields to add must also be deleted' );
    }

    $self->{del} = join (',', sort @del);

    if (!$self->{add} and !$self->{del} and !$self->error('del')) {
        $self->error( del => 'edit must not be empty' );
    }

    return $self;
}


sub error {
    my $self = shift;

    return (scalar keys %{$self->{errors}}) unless @_;
    
    my $attribute = shift;
    return $self->{errors}->{$attribute} unless @_;

    my $message = shift;
    $self->{errors}->{$attribute} = $message;

    return $message;
}


sub apply {
    my ($self, $pica, %args) = @_;
	my $strict = $args{strict};

    return if $self->error;

	if (!$pica) {
		$self->error( id => 'record not found' );
		return;
	} elsif ( $strict ) {
		if ( ($pica->ppn // '') ne $self->{ppn} ) {
			$self->error( id => 'PPN does not match' );
			return;
    	}

    	# TODO: check for disallowed fields to add/remove
	}

    my $iln = $self->{iln};
    my $epn = $self->{epn};

	# TODO: get ILN from record
	if ( $strict and $epn ne '' and $iln eq '' ) {
	    $self->error( iln => "ILN missing" );
		return;
	}

    my $add = PICA::Record->new( $self->{add} || '' );
    my $del = [ split ',', $self->{del} ];

    # new PICA record with all level0 fields but the ones to remove
    my @level0 = grep /^0/, @$del;
    my @level1 = grep /^1/, @$del;
    my @level2 = grep /^2/, @$del;

    # Level 0
    my $result = $pica->main;
    $result->remove( @level0 ) if @level0;
    $result->append( $add->main );    

    # Level 1
	if (@level1 or $add->field(qr/^1../)) {
		if ($strict and !$pica->holdings($iln)) {
			$self->error('iln', 'ILN not found');
			return;
		}
    }

    foreach my $h ( $pica->holdings ) {
        if ($iln and $iln eq ($h->iln // '')) {
            @level1 = map { $_ =~ qr{/} ? $_ : ($_,"$_/..") } @level1; 
            $h->remove( @level1 );
            $h->append( $add->field(qr/^1/) );
        } 
        $result->append( $h->fields );
	    # TODO: Level 2
    }
	
    $result->sort;

    return $result;
}


sub diff {
    my $self = shift;
    my $before = shift;

    my $after = $self->apply( @_ ) or return;
    require Text::Diff;

    my $l = scalar $before->fields + scalar $after->fields;
    my $diff = Text::Diff::diff(\$before,\$after,{CONTEXT => $l});
    $diff =~ s/^.+$//m;

    return $diff;
}

1;



__END__
=pod

=head1 NAME

PICA::Modification - Idempotent modifications of identified PICA+ records

=head1 VERSION

version 0.15

=head1 SYNOPSIS

  use PICA::Modification;

  # delete field '0123A' from record 'foo:ppn:123'
  my $mod = PICA::Modification->new( 
      id => 'foo:ppn:123', del => '0123A' 
  );

  $after = $mod->apply( $before );

=head1 DESCRIPTION

PICA::Modification models a modification of an identified PICA+ record
(L<PICA::Record>). The modification consist of the following attributes:

=over 4

=item add

A stringified PICA+ record with fields to be added.

=item del

A comma-separated list of PICA+ field tags to be removed. All tags of fields 
to be added must also be included for deletion so modifications are idempotent.

=item id

The fully qualified record identifier of form C<PREFIX:ppn:PPN>.

=item iln

The ILN of level 1 record to modify. Only required for modifications that
include level 1 fields.

=item epn

The EPN of the level 2 record to modify. Only required for modifications that
include level 2 fields.

=back

A modification instance may be malformed. A mapping from malformed attributes
to error messages is stored together with the PICA::Modification object.

PICA::Modification is extended to L<PICA::Modification::Request>. Collections
of modifications can be stored in a L<PICA::Modification::Queue>.

=head1 METHODS

=head2 check

Checks and normalized all attributes. A list of error messages is collected,
each connected to the attribute that an error originates from.

=head2 error( [ $attribute [ => $message ] ] )

Gets or sets an error message connected to an attribute. Without arguments this
method returns the number of errors.

=head2 apply ( $pica [, strict => 0|1 ] )

Applies the modification on a given PICA+ record and returns the resulting
record as L<PICA::Record> or C<undef> on malformed modifications. 

Only edits at level 0 and level 1 are supported by now.

The experimental argument C<strict> can be used to enable additional
validation. Validation errors are also collected in the PICA::Modification
object. A valid modification must:

=over 4

=item *

have a record identifier with PPN equal to the record's PPN (or both have none)

=item *

have ILN/EPN matching to the record's ILN/EPN (if given).

=back

=head2 diff

TODO: Test this!

=head2 new ( %attributes | {%attributes} | $modification )

Creates a new modification with given attributes. Missing attributes are set to
the empty string. On creation, all attributes are checked and normalized.

=head2 attributes

Returns a hash reference with attributes of this modification.

=head1 SEE ALSO

See L<PICA::Record> for information about PICA+ record format.

=encoding utf-8

=head1 AUTHOR

Jakob Voß <voss@gbv.de>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/PICA/Modification/Queue.pm
package PICA::Modification::Queue;
{
  $PICA::Modification::Queue::VERSION = '0.15';
}
#ABSTRACT: Queued list of modification requests of PICA+ records

use strict;
use warnings;
use v5.10;

use Carp;
use Scalar::Util qw(reftype);

sub new {
    my $class = shift;
    my $name  = shift || 'hash';
    
    if ('HASH' ~~ reftype($name)) {
        my $args = $name;
        $name = delete $name->{type};
        return $class->new( $name, %$args );
    }

    $class = $class . '::' . ucfirst($name);

    my $file = $class;
    $file =~ s!::!/!g;
    require "$file.pm"; ## no critic

    $class->new( @_ );
}

1;



__END__
=pod

=head1 NAME

PICA::Modification::Queue - Queued list of modification requests of PICA+ records

=head1 VERSION

version 0.15

=head1 SYNOPSIS

    # create a PICA::Modification::Queue::Hash
    my $q = PICA::Modification::Queue->new('Hash'); 

    # create a different kind of queue (all equivalent)
    $q = PICA::Modification::Queue->new('foo', bar => 'doz'); 
    $q = PICA::Modification::Queue->new({type => 'foo', bar => 'doz' }); 
    $q = PICA::Modification::Queue::Foo->new( bar => 'doz' ); 

    # operations on queues:
    
    my $id  = $q->insert( $modification );
    my $mod = $q->get( $id );
    $id     = $q->update( $id => $modification );
    $id     = $q->delete( $id );

    my @mods = $q->list( $key => $value ... );

=head1 DESCRIPTION

PICA::Modification::Queue implements a collection of modifications
(L<PICA::Modification>) or requests (L<PICA::Modification::Request>) on PICA+
records together with the CRUDL operations get, insert, update, delete, and
list. A queue may support insertion of modifications which then are stored as
modification requests.

The default implementation is a in-memory (non-persistent) hash
(L<PICA::Modification::Queue::Hash>). Additional types of queues can be
implemented in the C<PICA::Modification::Queue::> module namespace.

To test additional implementations of queues, the unit testing package 
L<PICA::Modification::TestQueue> should be used.

=head1 METHODS

=head2 new ( $name [, %options ] )

Creates a new queue. Options are passed to the queue's constructor.

=head2 request ( $modification )

Must return the id only on success.

=head2 get ( $id )

Returns a stored modification or undef.

=head2 update ( $id => $modification_or_modification_request )

Must return the id only on success.

=head2 delete ( $id )

Must return the id only on success.

=head2 list ( %parameters )

List all or a selection of queued modifications. Parameters can be used for
selection.  Special parameters are: C<page>, C<limit>, and C<sort>.

=encoding utf-8

=head1 AUTHOR

Jakob Voß <voss@gbv.de>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/PICA/Modification/Request.pm
package PICA::Modification::Request;
{
  $PICA::Modification::Request::VERSION = '0.15';
}
#ABSTRACT: Request for modification of an identified PICA+ record

use strict;
use warnings;
use v5.10;

use parent 'PICA::Modification';
use Time::Stamp gmstamp => { format => 'easy', tz => '' };

our @ATTRIBUTES = qw(id iln epn del add request creator status updated created);

sub new {
    my $self = PICA::Modification::new( @_ );
    
	$self->{created} //= gmstamp;
    $self->{status}  //= 0;

    $self;
}

sub update {
	my ($self, $status) = @_;

	$self->{status}  = $status;
	$self->{updated} = gmstamp;
}

1;



__END__
=pod

=head1 NAME

PICA::Modification::Request - Request for modification of an identified PICA+ record

=head1 VERSION

version 0.15

=head1 DESCRIPTION

PICA::Modification::Request extends L<PICA::Modification> with the following
attributes:

=over 4

=item request

Unique identifier of the request.

=item creator

Optional string to identify the creator of the request.

=item status

Status of the modification requests, which is 0 for unprocessed, 1 for
applied, and -1 for rejected.

=item created

Timestamp when the modification request was created (set automatically).

=item updated

Timestamp when the modification request was last updated (set automatically).

=back

All timestamps are GMT with format C<YYYY-MM-DD HH:MM::SS>.

=head1 METHODS

=head2 update ( $status )

Updates the status and sets the updated timestamp.

=encoding utf-8

=head1 AUTHOR

Jakob Voß <voss@gbv.de>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/PICA/Modification/TestQueue.pm
package PICA::Modification::TestQueue;
{
  $PICA::Modification::TestQueue::VERSION = '0.15';
}
#ABSTRACT: Unit test implementations of PICA::Modification::Queue

use strict;
use warnings;
use v5.10;

use Test::More 0.96;
use PICA::Modification;
use Test::JSON::Entails;

use parent 'Exporter';
our @EXPORT = qw(test_queue);

sub test_queue {
	my $queue = shift;
    my $name  = shift;

	subtest $name => sub {
	    my $test = bless { queue => $queue }, __PACKAGE__;
        $test->run;
    };
}

sub get { my $t = shift; $t->{queue}->get(@_); }
sub request { my $t = shift; $t->{queue}->request(@_); }
sub update { my $t = shift; $t->{queue}->update(@_); }
sub delete { my $t = shift; $t->{queue}->delete(@_); }
sub list { my $t = shift; $t->{queue}->list(@_); }

sub run {
	my $self = shift;

	my $list = $self->list();
	is_deeply $list, [], 'empty queue';

	my $mod = PICA::Modification->new( 
		del => '012A',
		id  => 'foo:ppn:123',
	);

	my $id = $self->request( $mod );
	ok( $id, "inserted modification" );

	my $got = $self->get($id);
	entails $got => $mod->attributes, 'get stored modification';
	isa_ok $got, 'PICA::Modification::Request';

	my $req = PICA::Modification::Request->new($mod);
    is $self->request($req), undef, 'reject insertion of modification requests';

	$list = $self->list();
	is scalar @$list, 1, 'list size 1';
	entails $list->[0] => $mod->attributes, 'list contains modification';

	$list = $self->list(limit => 1);
	is scalar @$list, 1, 'list option "limit"';

    $mod = PICA::Modification->new( del => '012A', id => 'bar:ppn:123' );
    my $id2 = $self->request( $mod );
	$list = $self->list( sort => 'id' );
	is scalar @$list, 2, 'list size 2 after inserting second modification';
    is $list->[0]->{id}, 'bar:ppn:123', 'list can be sorted';

	$list = $self->list( id => 'foo:ppn:123' );
	is scalar @$list, 1, 'search by field value';
    is $list->[0]->{id}, 'foo:ppn:123', 'only list matching modifications';

    foreach (0..4) {
        $mod = PICA::Modification->new( del => '012A', id => "doz:ppn:1$_" );
        $self->request($mod);
    }
    $list = $self->list( sort => 'id', limit => 3 );
    is scalar @$list, 3, 'inserted five additional modifications, limit works';

    $list = $self->list( sort => 'id', limit => 3, page => 2 );
    is scalar @$list, 3, 'limit';
    is $list->[0]->{id}, 'doz:ppn:12', 'page';

    $mod = PICA::Modification->new( add => '028A $xfoo', del => '028A', id => 'ab:ppn:1' );
    $id2 = $self->update( $id => $mod );
    is $id2, $id, 'update allowed';
    $mod = $self->get($id);
    is $mod->{del}, '028A', 'update changed';
    is $mod->{add}, '028A $xfoo', 'update changed';


	$mod = PICA::Modification->new( del => '012A', id  => 'xxx' );
    is $self->request($mod), undef, 'reject modification with error';

	$mod = PICA::Modification->new( del => '012A', id  => 'xxx' );
    is $self->update( $id => $mod), undef, 'reject modification with error';

	my $delid = $self->delete($id);
	is $delid, $id, 'deleted modification';

	$got = $self->get($id);
	is $got, undef, 'deleted modification returns undef';
}

1;


__END__
=pod

=head1 NAME

PICA::Modification::TestQueue - Unit test implementations of PICA::Modification::Queue

=head1 VERSION

version 0.15

=head1 SYNOPSIS

    use PICA::Modification::TestQueue;

    test_queue $queue, 'tested queue';

=head1 DESCRIPTION

This package exports the function C<test_queue> to run a simple unit test on a
L<PICA::Modification::Queue>.

=head1 AUTHOR

Jakob Voß <voss@gbv.de>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/PICA/Modification/Queue/Hash.pm
package PICA::Modification::Queue::Hash;
{
  $PICA::Modification::Queue::Hash::VERSION = '0.15';
}
#ABSTRACT: In-memory collection of modifications

use strict;
use warnings;
use v5.10;

use PICA::Modification::Request;
use Scalar::Util qw(blessed);

sub new { 
    bless [{},0], shift; 
}

sub get {
   my ($self, $id) = @_;
   return $self->[0]->{ $id };
}

sub request {
	my ($self, $mod) = @_;
	return if !$mod->isa('PICA::Modification') or $mod->isa('PICA::Modification::Request');
	$mod = PICA::Modification::Request->new( $mod );
	return if $mod->error;
	my $id = ++$self->[1];
	$self->[0]->{ $id } = $mod;
	return $id;
}

sub update { 
    my ($self, $id => $mod) = @_;
	$mod = PICA::Modification::Request->new( $mod );
	return if $mod->error;
	$self->[0]->{ $id } = $mod;
	return $id;
}

sub delete {
    my ($self, $id) = @_;
    return unless defined $self->[0]->{ $id };
	delete $self->[0]->{ $id }; 
	$id;
}

sub list {
	my ($self, %properties) = @_;

	my $limit  = delete $properties{limit} || 20;
	my $page   = delete $properties{page} || 1;
	my $sortby = delete $properties{sort};

	my $hash = $self->[0];

    my $c = 0;
    my $grep = sub {
        while (my ($k,$v) = each(%properties)) {
            return 0 unless $_[0]->{$k} eq $v; 
        }
        $c++;
        return 0 if $c > $page*$limit;
        return 0 if $c <= ($page-1)*$limit;
        1;
    };

    if ( $sortby ) {
        my $sort = sub { $a->{$sortby} cmp $b->{$sortby} };
        return [ grep { $grep->($_) } sort $sort values %$hash ];
    } else {
        return [ grep { $grep->($_) } values %$hash ];
    }
}

1;


__END__
=pod

=head1 NAME

PICA::Modification::Queue::Hash - In-memory collection of modifications

=head1 VERSION

version 0.15

=head1 DESCRIPTION

PICA::Modification::Queue::Hash implements a L<PICA::Modification::Queue> that
directly stored modification requests as references in memory.

=encoding utf-8

=head1 AUTHOR

Jakob Voß <voss@gbv.de>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/PICA/Modification/Queue/Smart.pm
package PICA::Modification::Queue::Smart;
{
  $PICA::Modification::Queue::Smart::VERSION = '0.15';
}
#ABSTRACT: Queued list of modification requests with checks

use strict;
use warnings;
use v5.10.1;

use parent 'PICA::Modification::Queue';

use Carp;
use Scalar::Util qw(reftype blessed);
use PICA::Record;
use LWP::Simple ();
use Time::Stamp gmstamp => { format => 'easy', tz => '' };

sub new {
    my $class = shift;
    my %args = @_ > 1 ? @_ : %{$_[0]};

    $args{queue} = PICA::Modification::Queue->new( $args{queue} )
        unless 'PICA::Modification::Queue' ~~ reftype $args{queue};
    
    $args{check} ||= 60;

    if (($args{via} // '') =~ /^https?:\/\//) {
        $args{unapi} = $args{via};
        $args{via} = sub {
            my $id = shift;
            return eval { 
                my $url = $args{unapi} . '?format=pp&id=' . $id;
                PICA::Record->new( LWP::Simple::get( $url ) ); 
            };
        };
    }

    croak "missing 'via' parameter to retrieve PICA+ records from"
        unless 'CODE' ~~ reftype $args{via};

    bless \%args, $class;
}

sub get {
    my ($self, $id) = @_;
    my $request = $self->{queue}->get($id) || return;

    return $request if $request->{status} != 0;

    # TODO: reject on error?

    my $last = $request->{updated} || $request->{created};
    my $next = gmstamp(time()-$self->{check});

    return $request if ($next cmp $last) == -1;

    $request->update( 0 ~~ $self->pending($request) ? 1 : 0 );
    $self->{queue}->update( $id => $request );
    $self->{queue}->get($id);
}

sub request {
    my ($self,$mod) = @_;

    my $modifies = $self->pending( $mod ) // return;

    if($modifies) {
        return $self->{queue}->request($mod);
    } else {
        my $id = $self->{queue}->request($mod) || return;
        my $request = $self->{queue}->get($id) || return;
        $request->{status} = 1;
        return $self->{queue}->update( $id => $request );
    }
}

sub update { my $self = shift; $self->{queue}->update(@_); }
sub delete { my $self = shift; $self->{queue}->delete(@_); }
sub list   { my $self = shift; $self->{queue}->list(@_); }


sub pending {
    my ($self, $mod) = @_;

    my $before = $self->{via}->( $mod->{id} );
    return unless blessed $before and $before->isa('PICA::Record');

    my $after = $mod->apply( $before, strict => 1 ) || return;

    return ($before->string eq $after->string ? 0 : 1);
}

1;



__END__
=pod

=head1 NAME

PICA::Modification::Queue::Smart - Queued list of modification requests with checks

=head1 VERSION

version 0.15

=head1 SYNOPSIS

  # wrap another queue, check after one minute (60 seconds) via unAPI at $url
  my $q = PICA::Modification::Queue::Smart->new( 
      queue => $queue, check => 60, via => $url 
  );

=head1 DESCRIPTION

PICA::Modification::Queue::Smart wraps another L<PICA::Modification::Queue> and
checks pending modification whether they have been applied:

=over

=item *

New requests are rejected unless the record to be modified could be retrieved.

=item *

New request resulting in no change are automatically set to status 1 (applied).

=item *

On get it is checked whether the modification has been applied after at least
C<check> seconds.

=back

=head1 METHODS

=head2 pending( $modification )

Checks whether a modification or modification request is still pending.
Returns 0, 1, or undef (if the status could not be checked).

=encoding utf-8

=head1 AUTHOR

Jakob Voß <voss@gbv.de>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Idempotent modifications of identified PICA+ records",
  "AUTHOR" => "Jakob Vo\303\237 <voss\@gbv.de>",
  "BUILD_REQUIRES" => {
    "Test::Exception" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "PICA-Modification",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "PICA::Modification",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Exporter" => 0,
    "LWP::Simple" => 0,
    "PICA::Record" => "0.584",
    "Scalar::Util" => 0,
    "Test::JSON::Entails" => 0,
    "Test::More" => "0.96",
    "Text::Diff" => 0,
    "Time::Stamp" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.15",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



